### PR TITLE
Gas limit beacon proposer preparer

### PR DESCRIPTION
**Pull request by mehdi-aouadi**
## PR Description

Adds gas limit parameter to the beacon proposer preparer

## Fixed Issue(s)

Preparation for https://github.com/ConsenSys/teku/issues/5999

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: validator/client/src/main/java/tech/pegasys/teku/validator/client/GasLimitProvider.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.client;
+
+import java.util.Optional;
+import tech.pegasys.teku.bls.BLSPublicKey;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public interface GasLimitProvider {
```
Comment (rolfyone) [inline validator/client/src/main/java/tech/pegasys/teku/validator/client/GasLimitProvider.java]: is this going to be needed separate to the FeeRecipientProvider, or should we be renaming that and adding to it? It seems like they're both just part of the same configuration?

File: validator/client/src/main/java/tech/pegasys/teku/validator/client/RuntimeProposerConfig.java [hunk 2]
```diff
@@ -68,20 +72,42 @@ public Optional<Eth1Address> getEth1AddressForPubKey(final BLSPublicKey pubKey)
...
-  public void delete(final BLSPublicKey publicKey) {
-    proposerConfigMap.remove(publicKey);
+  public synchronized void addOrUpdateGasLimit(
+      final BLSPublicKey publicKey, final UInt64 gasLimit) {
+    proposerConfigMap.put(
+        publicKey, new Config(getEth1AddressForPubKey(publicKey), Optional.ofNullable(gasLimit)));
     storagePath.ifPresent(this::save);
   }
 
+  public void deleteFeeRecipient(final BLSPublicKey publicKey) {
```
Comment (rolfyone) [inline validator/client/src/main/java/tech/pegasys/teku/validator/client/RuntimeProposerConfig.java]: given add and read are all synchronized, should probably make delete also synchronized..

File: validator/client/src/main/java/tech/pegasys/teku/validator/client/RuntimeProposerConfig.java [hunk 3]
```diff
@@ -68,20 +72,42 @@ public Optional<Eth1Address> getEth1AddressForPubKey(final BLSPublicKey pubKey)
...
+    if (proposerConfigMap.containsKey(publicKey)) {
+      addOrUpdateFeeRecipient(publicKey, null);
+      storagePath.ifPresent(this::save);
+    }
+  }
+
+  public void deleteGasLimit(final BLSPublicKey publicKey) {
```
Comment (rolfyone) [inline validator/client/src/main/java/tech/pegasys/teku/validator/client/RuntimeProposerConfig.java]: same here, synchronized

File: validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java [hunk 4]
```diff
@@ -229,6 +276,17 @@ private Optional<Eth1Address> getFeeRecipientFromProposerConfig(
     return config.getConfigForPubKey(publicKey).flatMap(Config::getFeeRecipient);
   }
 
+  private Optional<UInt64> getGasLimitFromProposerConfig(
+      final ProposerConfig config, final BLSPublicKey publicKey) {
+    return config
```
Comment (rolfyone) [inline validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java]: i think this should likely just be the same as above where we just return whatever is in config, and the providing of default value happens if this is empty... it looks like the get is already returning default, so im not sure why we'd provide default here...

File: validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java [hunk 5]
```diff
@@ -141,6 +147,24 @@ public Optional<Eth1Address> getFeeRecipient(final BLSPublicKey publicKey) {
...
+        .flatMap(config -> getGasLimitFromProposerConfig(config, publicKey))
+        .or(() -> runtimeProposerConfig.getGasLimitForPubKey(publicKey))
+        .or(
+            () ->
+                maybeProposerConfig.flatMap(
+                    proposerConfigProvider ->
+                        proposerConfigProvider.getDefaultConfig().getBuilder().isPresent()
+                            ? proposerConfigProvider.getDefaultConfig().getGasLimit()
+                            : Optional.empty()))
+        .or(() -> Optional.ofNullable(defaultGasLimit));
```
Comment (rolfyone) [inline validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java]: this should likely be a lot closer to the fee recipient also, flatmap should be handling empty so all the isPresent etc can disappear

**Review by rolfyone — COMMENTED** [review]
looks pretty close overall, a few bits and peices :)

File: validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java [hunk 6]
```diff
@@ -35,13 +35,15 @@
 import tech.pegasys.teku.validator.client.ProposerConfig.Config;
 import tech.pegasys.teku.validator.client.proposerconfig.ProposerConfigProvider;
 
-public class BeaconProposerPreparer implements ValidatorTimingChannel, FeeRecipientProvider {
+public class BeaconProposerPreparer
+    implements ValidatorTimingChannel, FeeRecipientProvider, GasLimitProvider {
```
Comment (rolfyone) [inline validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java]: probably already clear, but i think `FeeRecipeintProvider` and `GasLimitProvider` is probably a level of granularity too low, we might just want something like `ValidatorRegistrationDetailsProvider` or something...

Comment (mehdi-aouadi) [reply validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java]: It's all about finding a good generic name 😅

Comment (mehdi-aouadi) [reply validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java]: I named it `ValidatorRegistrationPropertiesProvider`

**Review by rolfyone — APPROVED** [review]
LGTM

**Pull request merged**